### PR TITLE
🌱 pkg/authorization: add delegation reason in audit

### PR DESCRIPTION
**Pull request by s-urbaniak**
## Summary
This is scavenging `DelegateAuthorization` from https://github.com/kcp-dev/kcp/pull/2089 to have delegation resons included in the kcp audit log.

Example for delegated authorization:
```
    "authorization.k8s.io/decision": "allow",
    "authorization.k8s.io/reason": "access granted",
    "bootstrap.authorization.kcp.dev/decision": "Allowed",
    "bootstrap.authorization.kcp.dev/reason": "bootstrap policy: RBAC: allowed by ClusterRoleBinding \"system:kcp:clusterworkspace:admin\" of ClusterRole \"cluster-admin\" to Group \"system:kcp:clusterworkspace:admin\"",
    "content.authorization.kcp.dev/decision": "Allowed",
    "content.authorization.kcp.dev/reason": "delegating due to \"root:e2e-org-b7zvn\" content policy: : systemcrd.authorization.kcp.dev: access granted",
    "maxpermissionpolicy.authorization.kcp.dev/decision": "Allowed",
    "maxpermissionpolicy.authorization.kcp.dev/reason": "delegating due to unbound resource: bootstrap.authorization.kcp.dev: access granted",
    "systemcrd.authorization.kcp.dev/decision": "Allowed",
    "systemcrd.authorization.kcp.dev/reason": "delegating due to no system CRD violation: maxpermissionpolicy.authorization.kcp.dev: access granted",
    "tenancy.kcp.dev/workspace": "root:e2e-org-b7zvn:provider-a",
    "toplevel.authorization.kcp.dev/decision": "Allowed",
    "toplevel.authorization.kcp.dev/reason": "delegating due to root workspace policy for \"e2e-org-b7zvn\": RBAC: allowed by ClusterRoleBinding \"system:admin:system:kcp:clusterworkspace:admin\" of ClusterRole \"cluster-admin\" to Group \"system:kcp:clusterworkspace:admin\": content.authorization.kcp.dev: access granted"
```

### CODE REVIEW

File: pkg/authorization/workspace_content_authorizer.go [hunk 1]
```diff
@@ -104,15 +104,15 @@ func (a *workspaceContentAuthorizer) Authorize(ctx context.Context, attr authori
...
 		}
-		return authorizer.DecisionNoOpinion, "root workspace access by non-root service account not permitted", nil
+		return authorizer.DecisionNoOpinion, "root workspace access by non-root service account", nil
 	}
 
 	// non-root workspaces must have a parent
 	parentClusterName, hasParent := cluster.Name.Parent()
 	if !hasParent {
-		return authorizer.DecisionNoOpinion, "non-root workspace that does not have a parent", nil
+		return authorizer.DecisionNoOpinion, fmt.Sprintf("parent less workspace"), nil
```
Comment (ncdc) [inline pkg/authorization/workspace_content_authorizer.go:115]:
```suggestion
		return authorizer.DecisionNoOpinion, fmt.Sprintf("parentless workspace"), nil
```

File: pkg/authorization/workspace_content_authorizer_test.go [hunk 2]
```diff
@@ -103,23 +103,23 @@ func TestWorkspaceContentAuthorizer(t *testing.T) {
 			requestedWorkspace: "root:unknown",
 			requestingUser:     newUser("user-1"),
 			wantDecision:       authorizer.DecisionDeny,
-			wantReason:         "clusterworkspace not found",
+			wantReason:         `workspace "root"|"unknown" not found`,
```
Comment (ncdc) [inline pkg/authorization/workspace_content_authorizer_test.go:106]: Do you want the quotes around path and name? Most/all of our other code like this does `%s|%s`

File: pkg/authorization/workspace_content_authorizer.go (hunk 1 above)
Comment (s-urbaniak) [reply pkg/authorization/workspace_content_authorizer.go:115]: oh thank you, fixed

File: pkg/authorization/workspace_content_authorizer_test.go (hunk 2 above)
Comment (s-urbaniak) [reply pkg/authorization/workspace_content_authorizer_test.go:106]: ah no, that's an artifact of using `%q|%q`, fixed.

File: pkg/authorization/decorator_test.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+	}{
+		"topAllows": {
+			authz: NewDecorator("top", alwaysAllow).AddAuditLogging().AddAnonymization().AddReasonAnnotation(),
+
+			wantDecision: authorizer.DecisionAllow,
+			wantReason:   "top: access granted",
+
+			wantAudit: map[string]string{
+				"top/decision": "Allowed",
+				"top/reason":   "unanonymized allow",
```
Comment (stevekuznetsov) [inline pkg/authorization/decorator_test.go:59]: If this is with anonymization, why is the reason unanonymyzed?

File: pkg/authorization/decorator_test.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+			},
+		},
+		"topAllowsWithoutReasonAnnotation": {
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/authorization/decorator_test.go:62]: so sorry, i was looking at prior art from other tests here so I wanted to stay idiomatic 😛

File: pkg/authorization/decorator_test.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+			authz: NewDecorator("top", alwaysAllow).AddAuditLogging().AddAnonymization(),
+
+			wantDecision: authorizer.DecisionAllow,
+			wantReason:   "access granted",
+
+			wantAudit: map[string]string{
+				"top/decision": "Allowed",
+				"top/reason":   "unanonymized allow",
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/authorization/decorator_test.go:70]: that's the audit entry, the difference is in the reason that bubbles up:

- `topAllows` (with `AddReasonAnnotation()`): `wantReason:   "top: access granted",`
- `topAllowsWithoutReasonAnnotation` (without `AddReasonAnnotation()`): `wantReason:   "access granted"`

The difference is the `top: ` in the first test that adds the reason annotation. I'll add a godoc comment to make it clear 👍

File: pkg/authorization/decorator_test.go (hunk 3 above)
Comment (s-urbaniak) [reply pkg/authorization/decorator_test.go:59]: Similar to `AddReasonAnnotation` anonmyzation refers to what reason bubbles up in the chain, in the audit we want to see the unanomyzied entry, let me add a godoc comment here as well.

File: pkg/authorization/decorator_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply pkg/authorization/decorator_test.go:70]: I see, I see. Sorry!

**Pull request merged**